src/chromo/models/__init__.py

```diff
@@ -35,3 +36,89 @@
     "Pythia6",
     "Pythia8",
 )
+
+
+# This is a workaround of a problem with "fixed" wheels
+# which is produced by "delocate" package on cibuildwheels
+# It changes Python -> the path of libpython in the "*.so" files
+# using standard "otool" and "install_name_tool" routines
+def fix_macos_installation(logfile):
+    import subprocess
+    import re
+    from pathlib import Path
+    from distutils.sysconfig import get_config_var
+
+    models_dir = Path(__file__).parent
+    for ext_file in models_dir.glob("*.so"):
+        ext_file = str(ext_file)
+
+        # Get libraries required by extension module
+        cmd = ["otool", "-L", ext_file]
+        result = subprocess.run(cmd, stdout=subprocess.PIPE)
+        lib_deps = result.stdout.decode("utf-8")
+
+        # Write to the logfile the list of libraries
+        # required by extension module before fixing
+        message = f"\n----------\nFixing {ext_file}:\nBefore\n{lib_deps}"
+        with open(logfile, "a") as lf:
+            lf.write(message)
+
```

Comment:
This function is very complex and should go into a separate module and only get imported into `__init__.py`

Reply:
Can you explain better why this is necessary? cibuildwheels works fine for me in iminuit and pyhepmc without this.

Reply:
Thinking more about it, this should not be run every time you include `models`. It should be run when the wheel is build. This is not the right place for such a correction. If you run it only once, during wheel building, you don't need to generate a log file to skip it if it has been run before.

Reply:
There was a problem on MacOS: when the chromo was installed from wheels, the code crashed with segmenation fault during loading extension modules. I found out it happened because of "Python" shared library. cibuildwheels launches `delocate` to repair wheels. Delocate goes through extension module and changes paths to external shared libraries to local path (.dylibs directory) and copy those external shared libraries to .dylibs directory.

For some reason on MacOS, `delocate` copies `Python` library to .dylibs. After installing from wheels  such "repaired" extension modules try to load that local copy of `Python` library. I don't know why this happens, but during of `Python` the program crashes with segmentation fault. I guess this happens because `python` loads its own copy of `Python` shared library. 

It is found, that if path to `Python` library in extension module change to the path on the system, i.e. to the `Python`  shared library located on the system which python loads at the start, the loading of extension module runs without problems. In this case only one copy of `Python` library is loaded.

The function changes the path with `Python` (.dylibs) in extension modules to the location of `Python` used by python which installs and lauches the code.

Ideally, this repair should be done during `pip install` command. But I don't know how to do it. Currently, the repairment runs during first initialization. Because it should be done only once, it checks for existence of the file "macos_fixed_ext.log". If the file is missing, the repair runs, otherwise it does nothing.

Reply:
Both iminuit and pyhepmc use cibuildwheels to build the wheels for Linux, MacOS, and Windows and I have not encountered that problem in these two projects. If you look into the build system of pyhepmc, it is very simple, there are no workarounds required.

I don't understand why this is required here.

Reply:
I downloaded the pyhepmc universal wheel for Macos pyhepmc-2.12.0-cp311-cp311-macosx_10_9_universal2.whl

You can unpack this with `tar xzf <filename>`

The module inside does not link to the Python library.
```
otool -L _core.cpython-311-darwin.so
_core.cpython-311-darwin.so:
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1300.36.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
```

Reply:
The `Python` dependency is probably from f2py. I tried to remove it, but it didn't work. You can check wheels for chromo on MacOS.

Reply:
I moved `fix_macos_installation` in util.py.

Reply:
Just to confirm, I ran
```
otool -L _sib21.cpython-310-darwin.so
_sib21.cpython-310-darwin.so:
	/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/Python (compatibility version 3.10.0, current version 3.10.0)
	/usr/local/opt/gcc/lib/gcc/current/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
	/usr/local/opt/gcc/lib/gcc/current/libgcc_s.1.1.dylib (compatibility version 1.0.0, current version 1.1.0)
	/usr/local/opt/gcc/lib/gcc/current/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
```
So we need to figure out one of the two things:
a) How to remove `/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/Python (compatibility version 3.10.0, current version 3.10.0)`
b) How to modify the files in the wheel on the build machine and not on the user machine.

Regarding b), you say the modification must happen during `pip install`. Why? I think you can let cibuildwheel make the wheel, and fix it with a script in the CI job before it is uploaded as an artifact. The script would decompress the wheel, and undo the act of delocate.

Reply:
It looks like a) is the real problem. I checked how it looks for numpy and scipy. Both are good role models for us, because they build multiple extensions, some from C/C++ code, some from Fortran code, and at least numpy uses f2py to sure.

Neither of these packages links to `/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/Python (compatibility version 3.10.0, current version 3.10.0)`

Reply:
The Pythia8 extension also does not link to libpython.

```
otool -L src/chromo/models/_pythia8.cpython-310-darwin.so
src/chromo/models/_pythia8.cpython-310-darwin.so:
	@rpath/libpythia8.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/gcc/lib/gcc/current/libstdc++.6.dylib (compatibility version 7.0.0, current version 7.30.0)
	/usr/local/opt/gcc/lib/gcc/current/libgcc_s.1.1.dylib (compatibility version 1.0.0, current version 1.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
```

Reply:
See #126 It seems to work locally.

Reply:
There is another problem on cibuildwheels. It is related to 

`Python_add_library(${target_name} MODULE WITH_SOABI` in F2PY.cmake.

For some reasons it add wrong suffix: all extension modules get `11` suffix, like `_sib21.cpython-311-darwin.so` even it is build for other versions of python.

Reply:
It is solved with "VIRTUAL_ENV" environment variable. See https://github.com/impy-project/chromo/pull/82#discussion_r1106672075